File: go/networkcore/protocol.go
// Package networkcore es el port a Go de forgenet/csharp/NetworkCore:
// transporte UDP genérico (handshake, heartbeat, input, snapshot, reliable,
// ping) sin ningún esquema de juego hardcodeado. El juego se engancha vía
// callbacks (OnInput, OnTick, StateProvider, QueueEvent) — ver host.go.
//
// Es la misma spec de protocolo binario que el core en C#: comparten
// framing (header, handshake, input, ack, ping) y, a diferencia de la
// versión previa de este servidor (que hardcodeaba PlayerState), ahora
// también comparten el formato de snapshot (Tick + StatePayload opaco +
// Events), porque los dos se generalizaron de la misma forma.
package networkcore

import "encoding/binary"

// Packet types
const (
	PacketSnapshot   = 0x01
	PacketInput      = 0x02
	PacketAck        = 0x03
	PacketPing       = 0x04
	PacketHandshake  = 0x05
	PacketDisconnect = 0x06
)

// Flags
const (
	FlagCompressed = 0x01
	FlagReliable   = 0x02
	FlagOrdered    = 0x04
)

// HeaderSize: seq(4) + ack(4) + typeAndFlags(1)
const HeaderSize = 9

func buildHeader(seq, ack uint32, packetType, flags byte) []byte {
	buf := make([]byte, HeaderSize)
	binary.BigEndian.PutUint32(buf[0:4], seq)
	binary.BigEndian.PutUint32(buf[4:8], ack)
	buf[8] = (packetType & 0x0F) | ((flags & 0x0F) << 4)
	return buf
}

func parseHeader(data []byte) (seq, ack uint32, packetType, flags byte, ok bool) {
	if len(data) < HeaderSize {
		return 0, 0, 0, 0, false
	}
	seq = binary.BigEndian.Uint32(data[0:4])
	ack = binary.BigEndian.Uint32(data[4:8])
	typeAndFlags := data[8]
	packetType = typeAndFlags & 0x0F
	flags = (typeAndFlags >> 4) & 0x0F
	return seq, ack, packetType, flags, true
}

// --- Payload del handshake (join) ---
//
// El core no sabe qué juego corre encima, así que Role viaja como un byte
// opaco: cada juego define sus propios valores y su propio significado (ej.
// taca-taca usa un valor para "barra" y otro para "tablero" — ver su
// ejemplo), igual que ya pasa con StatePayload y GameEvent.EventType.

// Modos de handshake — quién crea una sala nueva vs. quién se une a una
// existente por código. Esto sí es genérico: cualquier juego con sesiones
// de partida necesita esta distinción.
const (
	HandshakeModeCreate uint8 = 0x00
	HandshakeModeJoin   uint8 = 0x01
)

// Motivos de rechazo del handshake, enviados en un PacketDisconnect.
const (
	ReasonRoomNotFound uint8 = 0x01
	ReasonRoomFull     uint8 = 0x02
)

// EncodeJoinPayload: Mode(1) + Role(1) + RoomCodeLen(1) + RoomCode(N). Va
// después del header en un PacketHandshake enviado por el cliente. RoomCode
// se ignora en HandshakeModeCreate (el server genera el código).
func EncodeJoinPayload(mode, role uint8, roomCode string) []byte {
	buf := make([]byte, 0, 3+len(roomCode))
	buf = append(buf, mode, role, byte(len(roomCode)))
	buf = append(buf, roomCode...)
	return buf
}

func decodeJoinPayload(payload []byte) (mode, role uint8, roomCode string, ok bool) {
	if len(payload) < 3 {
		return 0, 0, "", false
	}
	codeLen := int(payload[2])
	if len(payload) < 3+codeLen {
		return 0, 0, "", false
	}
	return payload[0], payload[1], string(payload[3 : 3+codeLen]), true
}

// encodeHandshakeAck: PlayerID(2) + RoomCodeLen(1) + RoomCode(N). Va después
// del header en la respuesta del server — tanto quien crea la sala como
// quien se une reciben el código, para poder compartirlo (ej. un QR).
func encodeHandshakeAck(playerID uint16, roomCode string) []byte {
	buf := make([]byte, 0, 3+len(roomCode))
	buf = append(buf, byte(playerID>>8), byte(playerID), byte(len(roomCode)))
	buf = append(buf, roomCode...)
	return buf
}

File: rust/networkcore/src/protocol.rs
//! Constantes y helpers de bajo nivel del protocolo binario. Misma spec que
//! nucleo-multiplayer/go/networkcore y nucleo-multiplayer/csharp/NetworkCore
//! (mismo framing: header, handshake, input, ack, ping, y ahora también el
//! mismo formato de snapshot genérico con StatePayload opaco).

pub const PACKET_SNAPSHOT: u8 = 0x01;
pub const PACKET_INPUT: u8 = 0x02;
pub const PACKET_ACK: u8 = 0x03;
pub const PACKET_PING: u8 = 0x04;
pub const PACKET_HANDSHAKE: u8 = 0x05;
pub const PACKET_DISCONNECT: u8 = 0x06;

#[allow(dead_code)]
pub const FLAG_COMPRESSED: u8 = 0x01;
pub const FLAG_RELIABLE: u8 = 0x02;
#[allow(dead_code)]
pub const FLAG_ORDERED: u8 = 0x04;

/// seq(4) + ack(4) + typeAndFlags(1)
pub const HEADER_SIZE: usize = 9;

pub fn build_header(seq: u32, ack: u32, packet_type: u8, flags: u8) -> Vec<u8> {
    let mut buf = Vec::with_capacity(HEADER_SIZE);
    buf.extend_from_slice(&seq.to_be_bytes());
    buf.extend_from_slice(&ack.to_be_bytes());
    buf.push((packet_type & 0x0F) | ((flags & 0x0F) << 4));
    buf
}

pub fn parse_header(data: &[u8]) -> Option<(u32, u32, u8, u8)> {
    if data.len() < HEADER_SIZE {
        return None;
    }
    let seq = u32::from_be_bytes([data[0], data[1], data[2], data[3]]);
    let ack = u32::from_be_bytes([data[4], data[5], data[6], data[7]]);
    let type_and_flags = data[8];
    Some((seq, ack, type_and_flags & 0x0F, (type_and_flags >> 4) & 0x0F))
}

// --- Payload del handshake (join) ---
//
// El core no sabe qué juego corre encima, así que Role viaja como un byte
// opaco: cada juego define sus propios valores y su propio significado (ej.
// taca-taca usa un valor para "barra" y otro para "tablero" — ver su
// ejemplo), igual que ya pasa con StatePayload y GameEvent::event_type.

/// Modos de handshake — quién crea una sala nueva vs. quién se une a una
/// existente por código. Esto sí es genérico: cualquier juego con sesiones
/// de partida necesita esta distinción.
pub const HANDSHAKE_MODE_CREATE: u8 = 0x00;
pub const HANDSHAKE_MODE_JOIN: u8 = 0x01;

/// Motivos de rechazo del handshake, enviados en un PACKET_DISCONNECT.
pub const REASON_ROOM_NOT_FOUND: u8 = 0x01;
pub const REASON_ROOM_FULL: u8 = 0x02;

/// Mode(1) + Role(1) + RoomCodeLen(1) + RoomCode(N). Va después del header
/// en un PACKET_HANDSHAKE mandado por el cliente. RoomCode se ignora en
/// HANDSHAKE_MODE_CREATE (el server genera el código).
pub fn encode_join_payload(mode: u8, role: u8, room_code: &str) -> Vec<u8> {
    let mut buf = Vec::with_capacity(3 + room_code.len());
    buf.push(mode);
    buf.push(role);
    buf.push(room_code.len() as u8);
    buf.extend_from_slice(room_code.as_bytes());
    buf
}

pub fn decode_join_payload(payload: &[u8]) -> Option<(u8, u8, String)> {
    if payload.len() < 3 {
        return None;
    }
    let code_len = payload[2] as usize;
    if payload.len() < 3 + code_len {
        return None;
    }
    let room_code = String::from_utf8_lossy(&payload[3..3 + code_len]).into_owned();
    Some((payload[0], payload[1], room_code))
}

/// encode_handshake_ack: PlayerID(2) + RoomCodeLen(1) + RoomCode(N). Va
/// después del header en la respuesta del server — tanto quien crea la sala
/// como quien se une reciben el código, para poder compartirlo (ej. un QR).
pub fn encode_handshake_ack(player_id: u16, room_code: &str) -> Vec<u8> {
    let mut buf = Vec::with_capacity(3 + room_code.len());
    buf.extend_from_slice(&player_id.to_be_bytes());
    buf.push(room_code.len() as u8);
    buf.extend_from_slice(room_code.as_bytes());
    buf
}
